src/py/flwr/superexec/exec_servicer.py

```diff
@@ -56,13 +61,65 @@ def StartRun(
 
         self.runs[run.run_id] = run
 
+        # Start a background thread to capture the log output
+        capture_thread = threading.Thread(
+            target=_capture_logs, args=(run,), daemon=True
+        )
+        capture_thread.start()
+
         return StartRunResponse(run_id=run.run_id)
 
-    def StreamLogs(
+    def StreamLogs(  # pylint: disable=C0103
         self, request: StreamLogsRequest, context: grpc.ServicerContext
     ) -> Generator[StreamLogsResponse, Any, None]:
         """Get logs."""
-        logs = ["a", "b", "c"]
+        log(INFO, "ExecServicer.StreamLogs")
+
+        last_sent_index = 0
         while context.is_active():
-            for i in range(len(logs)):  # pylint: disable=C0200
-                yield StreamLogsResponse(log_output=logs[i])
+            # Exit if `run_id` not found
+            if request.run_id not in self.runs:
+                context.abort(grpc.StatusCode.NOT_FOUND, "Run ID not found")
+
+            # Yield n'th row of logs, if n'th row < len(logs)
+            logs = self.runs[request.run_id].logs
+            if last_sent_index < len(logs):
+                for i in range(last_sent_index, len(logs)):
+                    yield StreamLogsResponse(log_output=logs[i])
+                last_sent_index = len(logs)
+
+            # Shutdown context if process has completed. Previously stored
+            # logs will still be printed.
+            if self.runs[request.run_id].proc.poll() is not None:
+                log(INFO, "Run ID `%s` completed", request.run_id)
+                context.cancel()
+
+            time.sleep(0.1)  # Sleep briefly to avoid busy waiting
+
+
+def _capture_logs(
+    run: RunTracker,
+) -> None:
+    while not run.stop_event.is_set():
```

Comment:
Can we check if we need the `stop_event` here? I think we could do without if we use a `while True` as we later use `break` anyway.

Reply:
Checked and yes, it works using just `while True`. Much easier!